lib/midcom/helper/imagepopup/handler/upload.php

```diff
@@ -0,0 +1,155 @@
+        $mimetype = $temp['type'];
+        $parentguid = "";
+        
+        // Do we have guid ?
+        if($handler_id == '____ais-imagepopup-upload_image_noobject')
+        {
+            $parentguid = $this->_topic->guid;
+        }
+        else 
+        {
+            $parentguid = $args[0];
+        }
+        
+        // Set modified filename
+        $filename = $this->get_modify_filename($temp['name']);
+        
+        // Insert the image into database
+        $attachment = $this->insert_database($filename, $mimetype, $parentguid);
+        $data['attachment'] = $attachment;
+        
+        // Get target to write the file
+        $target = $this->get_data_from_database($filename, $parentguid);
+        
+        // Write the file
+        $this->write_the_file($temp_name, $target);
+        
+        // Make a response for editor.uploadImages() function
+        $location = midcom_db_attachment::get_url($attachment);
+        $data['location'] = $location;
+        
+        // Return image location as JSON
+        return new midcom_response_json(array('location' => $location));
+    }
+    
+    /**
+     * @param string $filename The file name
+     * @return string The modified file name
+     */
+    private function get_modify_filename($filename)
+    {
+        $filename = midcom_db_attachment::safe_filename($filename);
+        $pieces = explode('.', $filename);
+        $core = "modified";
+        $extension = array_pop($pieces);
+        $modifyFilename = $core . date('Y-m-d-H-i-s', time()) . "." . $extension;
+        
+        return $modifyFilename;
+    }
+    
+    /**
+     * @param string $filename The file name
+     * @param string $parentguid The parent GUID
+     * @return midcom_db_attachment The file destination
+     */
+    private function get_data_from_database($filename, $parentguid)
+    {
+        $query = midcom_db_attachment::new_query_builder();
+        $query->add_constraint('name', '=', $filename);
+        $query->add_constraint('parentguid', '=', $parentguid);
+        $entry = $query->execute();
+        
+        if(count($entry) == 0)
+        {
+            throw new midcom_error_notfound("There is no match in database " . midcom_connection::get_error_string());
+        }
+        else if(count($entry) == 1)
+        {
+            return $entry[0];
+        } 
+        else if(count($entry) > 1)
+        {
+            throw new midcom_error('There is more than just one object' . midcom_connection::get_error_string());
+        }
+    }
+    
+    /**
+     * @param string $tmp The temporary location of file
+     * @param midcom_db_attachment $target The final destination for file
+     */
+    private function write_the_file($tmp, midcom_db_attachment $target)
+    {
+        $source = fopen($tmp, 'r');
+        if(!$source)
+        {
+            throw new midcom_error("Could not open file " . $tmp . " for reading: " . midcom_connection::get_error_string());
```

Comment:
`midcom_connection::get_error_string` is not really useful here, because `fopen` is not a MidCOM function, but standard PHP